search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,433 @@
+{
+  EksArticleMetadataProvider *self = EKS_ARTICLE_METADATA_PROVIDER (object);
+
+  switch (prop_id)
+    {
+    case PROP_APPLICATION_ID:
+      g_clear_pointer (&self->application_id, g_free);
+      self->application_id = g_value_dup_string (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eks_article_metadata_provider_finalize (GObject *object)
+{
+  EksArticleMetadataProvider *self = EKS_ARTICLE_METADATA_PROVIDER (object);
+
+  g_clear_pointer (&self->application_id, g_free);
+  g_clear_object (&self->skeleton);
+  g_clear_object (&self->cancellable);
+
+  G_OBJECT_CLASS (eks_article_metadata_provider_parent_class)->finalize (object);
+}
+
+static void
+eks_article_metadata_provider_class_init (EksArticleMetadataProviderClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eks_article_metadata_provider_get_property;
+  object_class->set_property = eks_article_metadata_provider_set_property;
+  object_class->finalize = eks_article_metadata_provider_finalize;
+
+  eks_article_metadata_provider_props[PROP_APPLICATION_ID] =
+    g_param_spec_string ("application-id", "Application Id", "Application Id",
+      "", G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class,
+                                     NPROPS,
+                                     eks_article_metadata_provider_props);
+}
+
+typedef struct _ArticleMetadataQueryState {
+  EksArticleMetadataProvider *provider;
+  GDBusMethodInvocation      *invocation;
+} ArticleMetadataQueryState;
+
+static ArticleMetadataQueryState *
+article_metadata_query_state_new (EksArticleMetadataProvider *provider,
+                                  GDBusMethodInvocation      *invocation)
+{
+  ArticleMetadataQueryState *state = g_new0 (ArticleMetadataQueryState, 1);
+  state->provider = provider;
+  state->invocation = invocation;
+
+  return state;
+}
+
+static void
+article_metadata_query_state_free (ArticleMetadataQueryState *state)
+{
+  g_free (state);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ArticleMetadataQueryState,
+                               article_metadata_query_state_free);
+
+typedef struct GSList ObjectSList;
+
+static void
+object_slist_free (ObjectSList *slist)
+{
+  g_slist_free_full ((GSList *) slist, g_object_unref);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ObjectSList, object_slist_free);
+
+static void
+add_key_value_pair_to_variant (GVariantBuilder *builder,
+                               const char      *key,
+                               GVariant        *value)
+{
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("{sv}"));
+  g_variant_builder_add (builder, "s", key);
+  g_variant_builder_add (builder, "v", value);
+  g_variant_builder_close (builder);
+}
+
+static inline gchar *
+underscorify (const gchar *string)
```

Comment:
Is there no function that we can use in glib already that does this for property names, or does gobject not do this already for us when we use a property name with a dash but a property exists with an underscore in the name?

Reply:
I can try it to see if that works. At least to my recollection, there is no such function in GLib itself.

Reply:
Ah actually, the underscoring is for the benefit of the API consumer and not for GLib - the companion app service expects `everything to_come_like_snake_case` but the `property-names-use-dashes`